arduino-core/src/processing/app/helpers/StringReplacer.java

```diff
@@ -24,9 +24,50 @@
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.UUID;
 
 public class StringReplacer {
 
+  public static void checkIfRequiredKeyIsMissingOrExcept(String key, String src, PreferencesMap inDict) throws PreferencesMapException {
+    // If the key is not missing -> everything is OK
+    String checkedValue = inDict.get(key);
+    if (checkedValue != null && !checkedValue.isEmpty())
+      return;
+
+    PreferencesMap dict = new PreferencesMap(inDict);
+
+    // Find a random tag that is not contained in the dictionary and the src pattern
+    String tag;
+    while (true) {
+      tag = UUID.randomUUID().toString();
+      if (src.contains(tag))
+        continue;
+      if (dict.values().contains(tag))
+        continue;
+      if (dict.keySet().contains(tag))
+        continue;
+      break;
+    }
+
+    // Inject tag inside the dictionary
+    dict.put(key, tag);
+
+    // Recursive replace with a max depth of 10 levels.
+    String res;
+    for (int i = 0; i < 10; i++) {
+      // Do a replace with dictionary
+      res = StringReplacer.replaceFromMapping(src, dict);
+      if (res.equals(src))
+        break;
+      src = res;
+    }
+
+    // If the resulting string contains the tag, then the key is required
```

Comment:
This can't happen because `tag` is guaranteed to be different from any key/value contained in the map and either from any part of the `pattern` recipe, see how it is choosen in the `while` loop just before this.

Reply:
> Did you consider adding some instrumentation to replaceFromMapping to either check if a key is used, or more generically, let it return a list of keys it expanded?

This is trickier than it looks :-)

- returning the keys it expanded, won't help in detecting if a key is required, it could be that, for example, `{serial.port}` is required but it's not expanded because the key `serial.port` it's not present in the preferences map.
- we could "detect" a list of keys that needs to be replaced using a regexp that looks at matching braces in the recipes, but there are some recipes where the braces are actually needed! see this `openocd` command for instance:
`tools.openocd.program.pattern="{path}/{cmd}" {program.verbose} -s "{path}/share/openocd/scripts/" -f "{runtime.platform.path}/variants/{build.variant}/{build.openocdscript}" -c "telnet_port disabled; program {{{build.path}/{build.project_name}.elf}} verify reset; shutdown"`
the part `{{{build.path}/{build.project_name}.elf}}` will translate to `{{my/sketch.elf}}` but this doesn't mean that `{my/sketch.elf}` is a required field.